### PR TITLE
Reproduce attack and implement fix for liveness issue with ABA/commoncoin

### PR DESCRIPTION
**Pull request by sbellem**
Addresses amiller/HoneyBadgerBFT#59:

- [x] The constructive attack proposed by MacBrough should be reflected in the test cases
- [x] the ABA implementation must be modified to reflect the new fix

### NOTES
The adversarial network scheduler is currently limited to the minimal case of 4 nodes with one node being the attacker.

Modifying the network scheduler to support multiple nodes should not be too difficult, and is currently work in progress.

Logging was added for troubleshooting purposes. It is currently only meant to be used when testing, and future work will extend this work so that logging is available when the application is run in real scenarios.

### Reproducing the Attack
Run:

```bash
$ pytest -v -m demo test/demo_attack_issue59.py -s
```

### CODE REVIEW

**Comment by sbellem** [comment]
Note to myself (@sbellem): strange error on travis, will investigate why it is failing, and perhaps take the opportunity to use `docker` & `docker-compose` to run the tests on travis, because all tests pass on my machine, when using `docker`, so perhaps something is missing in the travis install step ...

File: honeybadgerbft/core/binaryagreement.py [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 from collections import defaultdict
+from distutils.util import strtobool
+from os import environ
+import logging
 
 from honeybadgerbft.exceptions import RedundantMessageError, AbandonedNodeError
 
 
+logger = logging.getLogger(__name__)
+CONF_PHASE = strtobool(environ.get('CONF_PHASE', '1'))
```
Comment (amiller) [inline honeybadgerbft/core/binaryagreement.py]: Are environment variables documented anywhere?

Comment (amiller) [reply honeybadgerbft/core/binaryagreement.py]: Is there any reason to disable the conf phase except for testing?

Comment (sbellem) [reply honeybadgerbft/core/binaryagreement.py]: I'll remove the switch. It was put there temporarily to ease the demonstration of the attack.

By putting the `CONF` phase logic in a function we could also simply mock that function in order to demonstrate the attack.

**Comment by sbellem** [comment]
@amiller The `CONF_PHASE` env var was removed and instead a separate "demo" test was written (commits 6842b7b4d030b0dadf969cdc65a2d4d68c7eb218 & 09fec408e1165ca177ff921570d38d29c416835d). 

This test should demonstrate the liveness issue, and consequently should only be invoked for demonstration purposes. Consequently invoking the test requires passing the option `-m demo` to pytest and also explicitly selecting the file. That is: 

```bash
pytest -v -m demo test/demo_attack_issue59.py -s
```

So, the following two alternatives won't work:

```bash
# file not picked up by default because it does not start with "test_"
pytest -v -m demo -s
```

```bash
# option "-m demo" missing
pytest -v test/demo_attack_issue59.py -s
```

**Comment by amiller** [comment]
This looks ready to go to me.

**Pull request merged**